### PR TITLE
Sign cookies #4656

### PR DESCRIPTION
**Pull request by SamMousa**
Sign cookies when the Yii2 application has cookie signing enabled

### CODE REVIEW

**Comment by SamMousa** [comment]
@samdark please check again, I have added a generic fix for CSRF.
It works by always adding a CSRF header; this should not result in any drawbacks for people not using it.

**Comment by samdark** [comment]
I don't really like latest commit. It practically equals to turning off CSRF validation. i.e. when you're manually submitting a form w/o a token it won't fail as expected.

**Comment by SamMousa** [comment]
Valid point, could you describe a syntax that you would like?
I can implement it!

**Comment by samdark** [comment]
I don't think any special syntax needed. CSRF token is generated and put into the form itself so submitting it via Codeception should cause the token to be submitted as well. We need to make sure that generated token is stored/read properly.

**Comment by SamMousa** [comment]
That's a good point, but for AJAX requests how would I handle it?
Or any other scenario where I don't want to test the form? 
Or originating page?

Something like `$I->setCsrfHeader()` or `$I->disableCsrf()`?

**Comment by samdark** [comment]
CSRF validation doesn't work for GET requests and is triggered for POST ones. POST could be:

1. Form submission. You have token in one of the fields.
2. AJAX request via `$I->sendAjaxPostRequest()`. We should not turn off validation there since real AJAX request will fail as well if submitted w/o a token. So there should be an easy way to obtain current token to pass it into `sendAjaxPostRequest()` parameters.

**Comment by SamMousa** [comment]
Okay I'll think on it and update tomorrow

**Comment by SamMousa** [comment]
Okay, I've been thinking on this for a bit.
Getting the token is not so easily done; to get it you need to instantiate the application and the request object. But before the request is sent the request object is re-instantiated.
The only way to force a specific token is to write it to the session or cookies, which could cause other side effects during testing.
At the very least it could cause confusion when random stuff starts showing up in the session / cookies during testing.

> So there should be an easy way to obtain current token to pass it into `sendAjaxPostRequest()` parameters.

For AJAX requests the normal approach (correct me if I'm wrong) is to send the CSRF via a header.
Obtaining the CSRF token from the server side and including (which is what you would be doing if you inject it into the AJAX params).
With that approach you are still not testing if the actual AJAX request generated by your application is correct or even includes the CSRF token.
When testing something using `sendAjaxPostRequest()` what I want to test is the handling of the data in the request. Therefore I propose the following syntax, of course function names can be changed:

````
$I->sendAjaxPostRequest(['controller/action']);
$I->seeResponseCodeIs(400); // CSRF error.

$I->disableCsrf();
$I->sendAjaxPostRequest(['controller/action']);
$I->seeResponseCodeIs(200);

$I->disableCsrf(); // Works for the 1 request only.
$I->sendAjaxPostRequest(['controller/action']);
$I->sendAjaxPostRequest(['controller/action']);
$I->seeResponseCodeIs(400); // CSRF error.
````

This approach allows me to easily test if CSRF is working correctly where it should be working, but it also allows me to disable it for a specific request.

What do you think?

**Comment by samdark** [comment]
Looks very artificial. Ideally it should work without `disableCsrf`.

**Comment by SamMousa** [comment]
> Looks very artificial. Ideally it should work without `disableCsrf`.
The alternative you proposed (correct me if I misunderstood) is;
````
$I->sendAjaxPostRequest(['controller/action']);
$I->seeResponseCodeIs(400); // CSRF error.

$csrfParam = $I->magicallyGetCsrfParamFromTheServer();
$csrfToken  = $I->magicallyGetCsrfTokenFromTheServer();
$I->sendAjaxPostRequest(['controller/action'], [
    $csrfParam => $csrfToken
]);
$I->seeResponseCodeIs(200);
````

Alternatively, if we force a page load before the ajax request, we could create a helper that reads the actual meta tag, that seems clean but incurs the extra request overhead:
````
$I->haveCsrfHeaderFromMetaTags();
$I->sendAjaxPostRequest(['controller/action']);
$I->seeResponseCodeIs(200);

// Or if you prefer to set it as data:
$I->sendAjaxPostRequest(['controller/action'], [
    $I->grabCsrfParamName() => $I->grabCsrfToken()
]);
$I->seeResponseCodeIs(200);
````

That could be quite clean, it could then extract the CSF param from the page using the HTML response parser (have to test if something like `grabTextContentFromXmlElement()` works for HTML. We have `seeElement()` so I can't imagine getting the text content of the element is hard.

**Comment by samdark** [comment]
I'm not saying we should pass it around with data. I'm saying that disabling it in order for it to pass is wrong because in a real app we don't have to. Tests should behave similar or else they're not precise.

**Comment by SamMousa** [comment]
I agree with that; but a test does not necessarily test all parts of the interaction.
Basically it's the same for logging in; you test it separately and then for the other tests you assume logging in / logging out works and use things like `amLoggedIn()`.

Anyway, do any of the other approaches I suggested seem acceptable?

**Comment by samdark** [comment]
Well, `magicallyGetCsrfParamFromTheServer` looks like the best approach to me.

**Comment by SamMousa** [comment]
It is also most complex, since there is no way to get it from the request and have it persist without setting cookies or session =/

**Comment by samdark** [comment]
Why avoiding setting session or cookies then?

**Comment by SamMousa** [comment]
Because it is a side effect.
````
// This would write to the session or cookie array.
$token = $I->getCsrfToken();
````

What about this:
````
// This would set the CSRF cookie to the value (and take care of masking); it is also clearly a setter.
$I->setCsrfCookie('abc');
// Do the actual test.
$I->sendAjaxPost('/', ['_csrf' => 'abc']);
````

**Comment by SamMousa** [comment]
Actual working code:
````
list($name, $masked) = $I->setCsrfCookie('abc');
        $I->sendAjaxPostRequest(['/mail-channel/target/remind-all', 'id' => $channel->id], [
            $name => $masked,
            'Mail' => [
                'en' => [
                    'from_email' => 'test'
                ]
            ]
        ]);
        
````

**Comment by samdark** [comment]
createAndSetCsrfCookie?

**Comment by SamMousa** [comment]
Okay, it has been implemented!
Code samples:

````
        // Use the REST module since we have no `sendAjaxDeleteRequest`.
        $I->haveHttpHeader('X-CSRF-Token', $I->setCsrfCookie('abc')[1]);
        $I->sendDELETE('/test');

        // Send CSRF as part of data:
       [$name, $val] = $I->setCsrfCookie('abc');
        $I->sendAjaxPostRequest('/test', [$name => $val]);
````

File: src/Codeception/Module/Yii2.php [hunk 1]
```diff
@@ -679,4 +679,26 @@ private function defineConstants()
         defined('YII_ENV') or define('YII_ENV', 'test');
         defined('YII_ENABLE_ERROR_HANDLER') or define('YII_ENABLE_ERROR_HANDLER', false);
     }
+
+    public function setCookie($name, $val, array $params = [])
```
Comment (samdark) [inline src/Codeception/Module/Yii2.php]: Need phpdoc.

**Comment by samdark** [comment]
@DavertMik I think it's ready. Please check.

File: src/Codeception/Module/Yii2.php (hunk 1 above)
Comment (samdark) [reply src/Codeception/Module/Yii2.php]: Yes.

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**